File: public/python/033-maximum-depth-of-binary-tree.py
# LeetCode Problem #104: Maximum Depth of Binary Tree
# Difficulty: Easy
# Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/

from typing import Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ─────────────────────────────────────────────
# APPROACH 1: Recursive DFS | O(n) time | O(h) space
# EXPLAIN: Depth = 1 + max(left_depth, right_depth); base case is None returning 0.
# WHEN: Cleanest solution; standard recursive pattern.

def maxDepth_recursive(root: Optional[TreeNode]) -> int:
    if not root:
        return 0
    return 1 + max(maxDepth_recursive(root.left), maxDepth_recursive(root.right))


# ─────────────────────────────────────────────
# APPROACH 2: BFS (Level Order) | O(n) time | O(w) space
# EXPLAIN: BFS counts levels; the total number of levels is the maximum depth.
# WHEN: Iterative approach; useful when recursion is disallowed.

def maxDepth_bfs(root: Optional[TreeNode]) -> int:
    if not root:
        return 0
    q = deque([root])
    depth = 0
    while q:
        depth += 1
        for _ in range(len(q)):
            node = q.popleft()
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
    return depth


# ─────────────────────────────────────────────
# APPROACH 3: Iterative DFS with Stack | O(n) time | O(h) space
# EXPLAIN: Use a stack of (node, current_depth) pairs and track the running maximum.
# WHEN: Iterative DFS; avoids Python recursion limit for very deep trees.

def maxDepth_stack(root: Optional[TreeNode]) -> int:
    if not root:
        return 0
    stack = [(root, 1)]
    max_depth = 0
    while stack:
        node, depth = stack.pop()
        max_depth = max(max_depth, depth)
        if node.left:
            stack.append((node.left, depth + 1))
        if node.right:
            stack.append((node.right, depth + 1))
    return max_depth


# ─────────────────────────────────────────────
# APPROACH 4: Morris Traversal | O(n) time | O(1) space
# EXPLAIN: Threaded-tree traversal tracks current depth without a stack or recursion.
# WHEN: Optimal space — O(1) extra; complex implementation, rarely asked.

def maxDepth_morris(root: Optional[TreeNode]) -> int:
    if not root:
        return 0
    max_depth = 0
    curr_depth = 0
    curr = root
    while curr:
        if not curr.left:
            curr_depth += 1
            max_depth = max(max_depth, curr_depth)
            curr = curr.right
        else:
            pred = curr.left
            steps = 1
            while pred.right and pred.right is not curr:
                pred = pred.right
                steps += 1
            if not pred.right:
                curr_depth += 1
                pred.right = curr
                curr = curr.left
            else:
                pred.right = None
                curr_depth -= steps
                curr = curr.right
    return max_depth


# ─────────────────────────────────────────────
# APPROACH 5: Tail-Recursion Style (Accumulator) | O(n) time | O(h) space
# EXPLAIN: Pass the current depth as an accumulator argument; update a mutable max on each leaf visit.
# WHEN: Demonstrates accumulator pattern; functionally same as Approach 1.

def maxDepth_accumulator(root: Optional[TreeNode]) -> int:
    result = [0]

    def helper(node, depth):
        if not node:
            return
        result[0] = max(result[0], depth)
        helper(node.left, depth + 1)
        helper(node.right, depth + 1)

    helper(root, 1)
    return result[0]


# Made with Bob

File: public/java/033-maximum-depth-of-binary-tree.java
/*
 * LeetCode Problem #104: Maximum Depth of Binary Tree
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // APPROACH 1: Recursive DFS | O(n) time | O(h) space
    // EXPLAIN: Depth = 1 + max(leftDepth, rightDepth); base case is null returning 0.
    public int maxDepth(TreeNode root) {
        if (root == null) return 0;
        return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }

    // APPROACH 2: BFS Level Order | O(n) time | O(w) space
    // EXPLAIN: BFS processes levels one at a time; the count of levels processed equals the maximum depth.
    public int maxDepthBFS(TreeNode root) {
        if (root == null) return 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        int depth = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            depth++;
            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                if (node.left  != null) q.offer(node.left);
                if (node.right != null) q.offer(node.right);
            }
        }
        return depth;
    }

    // APPROACH 3: Iterative DFS with Stack | O(n) time | O(h) space
    // EXPLAIN: Push (node, depth) pairs onto a stack; update maximum depth at each pop.
    public int maxDepthStack(TreeNode root) {
        if (root == null) return 0;
        Deque<Object[]> stack = new ArrayDeque<>();
        stack.push(new Object[]{root, 1});
        int maxDepth = 0;
        while (!stack.isEmpty()) {
            Object[] top = stack.pop();
            TreeNode node = (TreeNode) top[0];
            int depth = (int) top[1];
            maxDepth = Math.max(maxDepth, depth);
            if (node.left  != null) stack.push(new Object[]{node.left,  depth + 1});
            if (node.right != null) stack.push(new Object[]{node.right, depth + 1});
        }
        return maxDepth;
    }

    // APPROACH 4: Morris Traversal | O(n) time | O(1) space
    // EXPLAIN: Threaded-tree technique tracks current depth without recursion or stack; subtract steps when removing threads.
    public int maxDepthMorris(TreeNode root) {
        if (root == null) return 0;
        int maxDepth = 0, currDepth = 0;
        TreeNode curr = root;
        while (curr != null) {
            if (curr.left == null) {
                currDepth++;
                maxDepth = Math.max(maxDepth, currDepth);
                curr = curr.right;
            } else {
                TreeNode pred = curr.left;
                int steps = 1;
                while (pred.right != null && pred.right != curr) { pred = pred.right; steps++; }
                if (pred.right == null) {
                    currDepth++;
                    pred.right = curr;
                    curr = curr.left;
                } else {
                    pred.right = null;
                    currDepth -= steps;
                    curr = curr.right;
                }
            }
        }
        return maxDepth;
    }

    // APPROACH 5: Tail Recursion with Accumulator | O(n) time | O(h) space
    // EXPLAIN: Pass current depth as an accumulator; update a class-level max when traversing leaf branches.
    private int maxAcc = 0;

    public int maxDepthAccumulator(TreeNode root) {
        maxAcc = 0;
        helper(root, 1);
        return maxAcc;
    }

    private void helper(TreeNode node, int depth) {
        if (node == null) return;
        maxAcc = Math.max(maxAcc, depth);
        helper(node.left,  depth + 1);
        helper(node.right, depth + 1);
    }
}

// Made with Bob
